src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/BalloonPopup.java

```diff
@@ -164,26 +174,136 @@ public void recalculateBounds() {
         } else {
-          popupY = componentWindowPos.y - popupHeight - 5;
+          alignment = SwingConstants.TOP;
+          popupY = componentWindowPos().y - popupHeight() - 5;
         }
 
-        popupX = componentWindowPos.x + (anchorComponent.getWidth() - popupWidth) / 2;
+        popupX = componentWindowPos().x + (anchorComponent.getWidth() - popupWidth()) / 2;
+        outOfBoundsX = isOutOfBoundsX(0);
       }
 
       setTransparencyCoefficient(1.0f);
     }
 
-    setBounds(popupX, popupY, popupWidth, popupHeight);
+    var popupPoint = calculatePoint(popupX, popupY, oldAlignment, outOfBoundsX, outOfBoundsY);
+
+    popupX = popupPoint.x;
+    popupY = popupPoint.y;
+
+    setBounds(popupX, popupY, popupWidth(), popupHeight());
+  }
+
+  private boolean isOutOfBoundsX(int margin) {
+    return position != null
+        && isOutOfBounds(position.x, componentWindowPos().x, margin, popupWidth(), componentSize().width);
+  }
+
+  private boolean isOutOfBoundsY(int margin) {
+    return position != null
+        && isOutOfBounds(position.y, componentWindowPos().y, margin, popupHeight(), componentSize().height);
+  }
+
+  private boolean isOutOfBounds(int positionCoord, int componentWindowPos,
+                                int margin, int popupSize,
+                                int componentSize) {
+    return positionCoord < componentWindowPos + margin
+        || positionCoord + popupSize > componentWindowPos + componentSize - margin;
+  }
+
+  private int clampX(int x) {
+    return clamp(componentWindowPos().x, componentSize().width, popupWidth(), x);
+  }
+
+  private int clampY(int y) {
+    return clamp(componentWindowPos().y, componentSize().height, popupHeight(), y);
+  }
+
+  private int clamp(int componentWindowPos, int componentSize, int popupSize, int coord) {
+    int margin = positionCenter ? POPUP_MARGIN : 0;
+    return Integer.max(componentWindowPos + margin,
+        Integer.min(componentWindowPos + componentSize - popupSize - margin,
+            coord));
+  }
+
+  private Point calculatePoint(int popupX, int popupY, int oldAlignment, boolean outOfBoundsX, boolean outOfBoundsY) {
+    positionHorizontally = alignment == SwingConstants.LEFT || alignment == SwingConstants.RIGHT;
+    positionCenter = alignment == SwingConstants.CENTER;
+    var point = new Point(popupX, popupY);
+    if (oldAlignment == alignment && position != null) {
+      var positionX = position.x;
+      if (outOfBoundsX) {
+        positionX = clampX(positionX);
+      }
+      var positionY = position.y;
+      if (outOfBoundsY) {
+        positionY = clampY(positionY);
+      }
+      position.setLocation(positionX, positionY);
+      point.setLocation(positionCenter || !positionHorizontally ? positionX : popupX,
+          positionCenter || positionHorizontally ? positionY : popupY);
+    }
+    return point;
+  }
+
+  private Dimension componentSize() {
+    return anchorComponent.getSize();
```

Comment:
Was it this PR that were perhaps involved in some excessive memory use? This may be the cause. Consider either using `getWidth()` and `getHeight()` (see [docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.desktop/java/awt/Component.html#getWidth())).